src/utils.rs

```diff
@@ -56,6 +59,112 @@ pub fn validate_string(s: &str) -> bool {
     !s.is_empty()
 }
 
+#[cfg(test)]
+mod test {
+    #[allow(unused_imports)]
+    use super::*;
+
+    #[test]
+    fn slice_at_the_end() {
+        let s = "this is a test"; // No Unicode
+        assert_eq!(slice_utf8(s, 3, s.len()), &s[3..(s.len())]);
+    }
+
+    #[test]
+    fn slice_in_the_utf8() {
+        let s = "ϵthiϕś αβ a test"; // Unicode
+        assert_eq!(slice_utf8(s, 2, 6), "hiϕś");
+    }
+
+    #[test]
+    fn short() {
+        assert_eq!(slice_utf8("ö", 0, 1), "ö");
+    }
+
+    #[test]
+    fn arabic() {
+        let s = "من"; // entire string is unicode
+        assert_eq!(slice_utf8(s, 0, 2), s);
+    }
+
+    #[test]
+    fn empty() {
+        let s = "abcd";
+        assert_eq!(slice_utf8(s, 2, 2), &s[2..2]);
+    }
+}
+
+/// A function to handle slicing into strings which may contain unicode.
+///
+/// Unlike Python, Rust cares deeply about unicode strings and ensuring every
+/// slice into them remains valid. As slicing assumes that each Unicode
+/// character is either in or outside of a slice, a simple slice operation can
+/// panic when run on a Rust string. We manually slice a Rust string by
+/// iterating over it's character points. This avoids panic on unicode strings.
+///
+/// typical usage would be:
+///     `slice_utf8(simple_string, 3, 7)` instead of `&simple_string[3..7]``
+///
+/// Panics if `low` > `high` or `high` > `string.len`
+fn slice_utf8(string: &str, low: usize, high: usize) -> &str {
+    debug_assert!(low <= high);
+    debug_assert!(high <= string.len());
+    if low == high {
+        return "";
+    }
+    let mut indices = string.char_indices().enumerate().map(|(e, (i, _))| (i, e));
+    let mut low_index = None;
+    let mut high_index = None;
+    let mut calc_high_index = |mut indices: std::iter::Map<_, _>| {
+        high_index = Some(if let Some((i, _)) = indices.next() {
+            i
+        } else {
+            string.len()
+        })
+    };
+    // Find low bound
+    for (i, e) in &mut indices {
+        if e == low {
+            low_index = Some(i);
+            break;
+        }
+    }
+    // find high bound
+    if low + 1 == high {
```

Comment:
Perhaps a more canonical way to write this would be something like

```
let mut cursor = low + 1;
while (cursor != high) {
    cursor = indices.next(); // i think the types are wrong on this line, but you get the idea - break on None, etc.
}
```

it makes it more obvious, to me, that we're skimming through the indices to find the high point.

Reply:
We can do this, but we need to leave in the `if low + 1 == high {`. This because of how rust allows us to index one past the last character of an array. I cleaned up the syntax though.